ultravox/tools/enrich_synth_data.py

```diff
@@ -0,0 +1,101 @@
+
+Answer: {answer}
+
+Provide a short explanation to the question given the passage that entails the answer."""
+
+CLIENT = openai.Client()
+
+
+# This script is used to embellish a dataset with synthetic data generated by a language model.
+# Ex: just enrich_ds -d fixie-ai/boolq-audio -s train -b with_explanation -c explanation
+@dataclasses.dataclass
+class EnrichWithSyntheticDataArgs:
+    dataset_name: str = simple_parsing.field(alias="-d")
+    dataset_subset: str = simple_parsing.field(default="default", alias="-S")
+    dataset_split: Optional[str] = simple_parsing.field(default=None, alias="-s")
+    new_column_name: str = simple_parsing.field(default="explanation", alias="-c")
+    language_model: str = simple_parsing.field(default="gpt-4o", alias="-m")
+    template: str = DEFAULT_TEMPLATE
+    num_samples: Optional[int] = simple_parsing.field(default=None, alias="-n")
+    num_workers: int = simple_parsing.field(default=16, alias="-w")
+    upload_name: Optional[str] = simple_parsing.field(default=None, alias="-u")
+    upload_branch: Optional[str] = simple_parsing.field(default="main", alias="-b")
+    token: Optional[str] = simple_parsing.field(default=None, alias="-t")
+
+    def __post_init__(self):
+        if not self.upload_name:
+            self.upload_name = self.dataset_name
+
+
+def _enrich_sample(sample, template: str, new_col_name: str, language_model: str):
+    input_text = template.format(**sample)
+    response = CLIENT.chat.completions.create(
+        model=language_model,
+        messages=[{"role": "user", "content": input_text}],
+        max_tokens=128,
+        temperature=0,
+    )
+    sample[new_col_name] = response.choices[0].message.content
+    return sample
+
+
+def main(args: EnrichWithSyntheticDataArgs):
+    ds_name = args.dataset_name
+    new_col_name = args.new_column_name
+
+    print(
+        f'Loading dataset "{ds_name}", new column name: "{new_col_name}", template:\n{args.template}'
+    )
+    data_dict = datasets.load_dataset(
+        ds_name, args.dataset_subset, split=args.dataset_split
+    )
+    if args.dataset_split:
+        data_dict = {args.dataset_split: data_dict}
+    for split, ds_split in data_dict.items():
+        print(f'Processing split "{split}"...')
+        if args.num_samples:
+            ds_split = ds_split.select(range(args.num_samples))
+
+        new_split = ds_split.map(
+            _enrich_sample,
+            num_proc=args.num_workers,
+            fn_kwargs={
+                "language_model": args.language_model,
+                "new_col_name": new_col_name,
+                "template": args.template,
+            },
+        )
+
+        token = args.token or os.environ.get("HF_TOKEN")
+
+        try:
+            new_split.push_to_hub(
+                args.upload_name,
+                config_name=args.dataset_subset,
+                split=split,
+                token=token,
+                revision=args.upload_branch,
+            )
```

Comment:
This kept failing for me. I think the right approach is to apply this to all splits and upload the final `data_dict` but I didn't try it.

Reply:
Do you have a HF write token?

Reply:
Yes I do. The issue was that the schema was changing and it didn't allow changing the schema one-split at a time.

Reply:
ah ok, makes sense.

Reply:
I didn't verify if this works, but I'm now uploading the both splits at the same time. I think that would fix the issue since the splits won't be out of sync at any point in time.